### PR TITLE
Upload CSVs

### PR DESCRIPTION
**Pull request by ghickman**
This switches CSV importing from the CLI to a web form.

The importing itself it done by a Queue worker since it takes over 10 seconds.

The uploaded files are stored on the Version created when the form is submitted and we store a processed timestamp so we can infer the created but unprocessed state.

### CODE REVIEW

File: ncdr/importers/grouping.py [hunk 1]
```diff
@@ -1,24 +1,30 @@
...
-@transaction.atomic
-def load_file(file_name, version):
-    with open(file_name, "r", encoding="Windows-1252") as f:
-        f.readline()  # ignore the first line since it's blank
-        rows = list(csv.DictReader(f, delimiter="¬"))
+def load_file(fd, version):
+    fd.readline()  # ignore the first line since it's blank
+    rows = list(csv.DictReader(fd, delimiter="¬"))
+
+    existing_names = {g.name: g.description for g in Grouping.objects.all()}
```
Comment (fredkingham) [inline ncdr/importers/grouping.py]: It looks like we're only using the name from the existing names. Can we not do

```
set(Grouping.objects.all().values_list('name', flat=True))
```

Without trying it, I'm assuming this is faster and simpler. Note "Without trying it"!

File: ncdr/static/css/styles.scss [hunk 2]
```diff
@@ -657,3 +657,21 @@ hr {
...
+// Version upload form
+form.upload {
+  .form-group {
+    align-items: baseline;
+    display: flex;
+    margin-bottom: 0;
+  }
+
+  input[type="file"] {
+    display: flex;
```
Comment (fredkingham) [inline ncdr/static/css/styles.scss:676]: Have we checked on IE? Browser support for flex is not 100% but I'm not sure how limited.

File: ncdr/importers/column.py [hunk 3]
```diff
@@ -1,11 +1,34 @@
...
+    the time of writing, was called by an RQ worker it ran in ~10s.  After much
+    digging it was discovered that the aggregate time cost for the
+    get_or_creates calls was the cause, again only when called directly from
+    the view.  Rather than waste more time building an extension on the rabbit
+    hole the sensible decision to use a precomputed lookup table was taken.
+
+    This function builds that table, creating any missing DataElements in the
+    process.
+    """
+    existing_de_names = set(DataElement.objects.values_list("name", flat=True))
```
Comment (fredkingham) [inline ncdr/importers/column.py:26]: we could do existing_de_names = set(DataElement.objects.all().filter(name__in= csv_de_names).values_list("name", flat=True))

(bit bike sheddy...)

**Comment by fredkingham** [comment]
Should we add pre-commit to `requirments-dev.txt`?

**Comment by fredkingham** [comment]
Can we remove load date from the readme

**Comment by fredkingham** [comment]
Don't we need to add the redis install to ansible?

**Comment by fredkingham** [comment]
Can we create an initial version?

I tried migrating existing data and migrations blew up. I nuked the db and created and I get 

```

Version matching query does not exist.
--

```

(I could be being stupid!)

File: ncdr/static/css/styles.scss (hunk 2 above)
Comment (ghickman) [reply ncdr/static/css/styles.scss:676]: Sort of? https://caniuse.com/#feat=flexbox

Do you have an easy way to test this?

File: ncdr/importers/column.py (hunk 3 above)
Comment (ghickman) [reply ncdr/importers/column.py:26]: You still have to compute `missing_de_names` though so you've asked the database to do a slightly harder query (by adding a filter clause) without saving the CPU any time when it calculates the set difference.

File: ncdr/importers/grouping.py (hunk 1 above)
Comment (ghickman) [reply ncdr/importers/grouping.py]: Oops, that's some left over from an unnecessary change, good catch!  Fixed and rebased.

**Comment by ghickman** [comment]
> Should we add pre-commit to requirments-dev.txt?

oh go on then ;) 

> Can we remove load date from the readme

Done and rebased.


> Don't we need to add the redis install to ansible?

Yes, yes we do.  Added and rebased in.

**Comment by ghickman** [comment]
@fredkingham – updated to add an initial version and fix some errors I found when running from a clean database.  You should be able to review locally now!

**Comment by fredkingham** [comment]
When we're logged in can we have a button in the logged in bar to take you to `/upload`

**Comment by ghickman** [comment]
@fredkingham – sure.  Should it take you to `/upload` or `/versions` where upload is also linked?

**Comment by fredkingham** [comment]
@ghickman `/version`s is a better idea! (I chose looked for the url that would let me upload 😃 )

**Comment by fredkingham** [comment]
I've done my upload

Screen Shot 2019-03-12 at 16 32 26

**Comment by fredkingham** [comment]
Screen Shot 2019-03-13 at 10 41 18

We're missing a distinct somewhere maybe?

Steps to reproduce, upload files. Reupload files with a table name slightly changed.
Search for the database that contains that table.

We now see two databases in the matches, but both lead to the same url.

**Comment by fredkingham** [comment]
I'm getting a 500 when I try and publish the new version 
Screen Shot 2019-03-13 at 11 42 03

**Comment by ghickman** [comment]
@fredkingham – fixed the publishing bug, forgot I'd built `Version.[un]publish()` to handle audit logs already!  I've also fixed a bug with auth URL redirects and added that versions link to the logged in header.

**Comment by fredkingham** [comment]
I'm getting errors in the queue

`Version matching query does not exist`

```
NCDR-reference-library/ncdr/importers/__init__.py", line 32, in import_data
    version = Version.objects.get(pk=version_pk)
```

**Comment by fredkingham** [comment]
I'm still blowing up on version upload with Version does not exist. 

Initially it was because I had multiple versions published (I think). Now I'm not sure...

**Comment by fredkingham** [comment]
👍

**Pull request merged**